### PR TITLE
Add Root and Game classes to main logic

### PR DESCRIPTION
**Pull request by Sense101**

### CODE REVIEW

File: server/classes/player.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		/** @type {string} */
+		this.playerSecret = Math.random().toString()
+
+		/** @type {Array<string>} */
+		this.playerDeck = getStarterPack()
+
+		/** @type {string} */
+		this.playerName = playerName
+
+		/** @type {import('socket.io-client').Socket} */
```
Comment (martinkadlec0) [inline server/classes/player.js:26]: This is no client Socket, but backend socket.

File: server/classes/player.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+//import { Socket } from "socket.io-client"
```
Comment (martinkadlec0) [inline server/classes/player.js:1]: remove

File: server/classes/player.js (hunk 1 above)
Comment (martinkadlec0) [inline server/classes/player.js:26]: This is BE instance not client instance

File: server/classes/root.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/** The root of the server. All data stored on the server should be (@TODO) accessible from this object */
+export class Root {
+	constructor() {
+		/** @type {Object.<string, import('./player').Player>} */
```
Comment (martinkadlec0) [inline server/classes/root.js:4]: Perhaps to make things bit more tidy, we could import types at the top of the file?
```javascript
/** @typedef {import("./player")} Player*/
```

File: server/routines/chat.js [hunk 4]
```diff
@@ -1,35 +1,31 @@
 import {takeEvery, all} from 'redux-saga/effects'
-import gameSaga from './game'
+import {broadcast} from '../utils'
+import {Game} from '../classes/game'
```
Comment (martinkadlec0) [inline server/routines/chat.js:3]: import as type

File: server/routines/game.js [hunk 5]
```diff
@@ -22,6 +22,8 @@ import followUpSaga from './turn-actions/follow-up'
 import {HookMap, SyncHook, SyncBailHook, SyncWaterfallHook} from 'tapable'
 import registerCards from '../cards/card-plugins'
 import chatSaga from './chat'
+import {Game} from '../classes/game'
```
Comment (martinkadlec0) [inline server/routines/game.js:25]: import as types

File: server/routines/game.js [hunk 6]
```diff
@@ -253,47 +265,54 @@ function* turnActionSaga(game, turnAction, baseDerivedState) {
...
-	const opponentPlayer = game.state.players[opponentPlayerId]
+	const currentPlayerState = game.state.players[currentPlayerId]
+	const opponentPlayerState = game.state.players[opponentPlayerId]
 
-	game.state.turnPlayerId = currentPlayer.id
-
-	// console.log('NEW TURN: ', {currentPlayerId, opponentPlayerId})
+	game.state.turnPlayerId = currentPlayerId
 
+	// @TODO we can use game.state for everything instead now, but we will need to rewrite all cards for that first
```
Comment (martinkadlec0) [inline server/routines/game.js:305]: As discussed, this is not necessarily true.

File: server/routines/player.js [hunk 7]
```diff
@@ -1,15 +1,19 @@
-import {takeEvery, all, put, take, race, delay} from 'redux-saga/effects'
-import {getStarterPack} from '../utils/state-gen'
+import {takeEvery, put, take, race, delay} from 'redux-saga/effects'
 import {validateDeck} from '../utils'
 import CARDS from '../cards'
+import {Root} from '../classes/root'
```
Comment (martinkadlec0) [inline server/routines/player.js:4]: import as type

File: server/utils/index.js [hunk 8]
```diff
@@ -164,3 +165,18 @@ export const validateDeck = (deckCards) => {
...
 	if (deckCards.length !== 42) return 'Deck must have exactly 42 cards.'
 }
+
+/**
+ *
+ * @param {Array<Player>} players
+ * @param {string} type
+ * @param {*} payload
+ */
+export function broadcast(players, type, payload = {}) {
```
Comment (martinkadlec0) [inline server/utils/index.js:175]: I'd put this into its own utils file, something like `utils/socket` or `utils/messaging` or something like that.

**Review by martinkadlec0 — CHANGES_REQUESTED** [review]

File: server/classes/player.js (hunk 2 above)
Comment (Sense101) [reply server/classes/player.js:1]: yep

File: server/classes/player.js (hunk 1 above)
Comment (Sense101) [reply server/classes/player.js:26]: I know, is there a problem with the code there?

File: server/classes/root.js (hunk 3 above)
Comment (Sense101) [reply server/classes/root.js:4]: good idea

Comment (Sense101) [reply server/classes/root.js]: done for all files

**Pull request merged**